### PR TITLE
fix bug of celoss when using ignore_index and reduction

### PR DESCRIPTION
**Pull request by chajchaj**
### PR types
Bug fixes

### PR changes
APIs

### Describe
fix bug of using ignore_index and reduction

### CODE REVIEW

File: python/paddle/nn/functional/loss.py [hunk 1]
```diff
@@ -1223,22 +1223,44 @@ def cross_entropy(input,
...
-                return out_sum / total_weight
+            #1. if weight==none, 
+            #    numerator: reduce_sum all loss directly is ok causeof softmax_with_cross_entropy's inner logic
+            #    denominator: count sample num with class_index!=ignore_index
+            #2. else
+            #    numerator: loss's weighted sum 
+            #    denominator: cal the sum of weight where the sample's class_index!=ignore_index
+            out_sum = core.ops.reduce_sum(out, 'reduce_all', True)
+            #for each label[i],set 1 or 0, according to ignore_index
+            ones_t = paddle.ones_like(label)
```
Comment (XiaoguangHu01) [inline python/paddle/nn/functional/loss.py]: 这里判断下ignore_index这个设置是否生效，大部分场景是不需要使用ignore_index的吧
避免不需要ignore_index的地方，造成性能下降。

File: python/paddle/nn/functional/loss.py [hunk 2]
```diff
@@ -1258,20 +1280,33 @@ def cross_entropy(input,
...
     if reduction == "sum":
         return paddle.sum(out, name=name)
     elif reduction == "mean":
-        if weight is not None:
-            out_sum = paddle.sum(out, name=name)
-            total_weight = paddle.sum(weight_gather_reshape)
-            return out_sum / total_weight
+        out_sum = paddle.sum(out, name=name)
+        #for each label[i],set 1 or 0, according to ignore_index
+        ones_t = paddle.ones_like(label)
```
Comment (XiaoguangHu01) [inline python/paddle/nn/functional/loss.py]: 同上

**Review by XiaoguangHu01 — DISMISSED** [review]
LGTM

File: python/paddle/nn/functional/loss.py [hunk 3]
```diff
@@ -1223,22 +1223,53 @@ def cross_entropy(input,
...
+            #    numerator: loss's weighted sum 
+            #    denominator: cal the sum of weight where the sample's class_index!=ignore_index
+            if ignore_index != -100:
+                out_sum = core.ops.reduce_sum(out, 'reduce_all', True)
+                #for each label[i],set 1 or 0, according to ignore_index
+                ones_t = paddle.ones_like(label)
+                zeros_t = paddle.zeros_like(label)
+                #mask[i]=0, if label[i]==ignore_index
+                #mask[i]=1, otherwise 
+                mask = paddle.where(label != ignore_index, ones_t, zeros_t)
```
Comment (willthefrog) [inline python/paddle/nn/functional/loss.py]: maybe simplify a bit with
```python
paddle.sum(paddle.cast(label != ignore_index, dtype='float32'))
```

File: python/paddle/nn/functional/loss.py [hunk 4]
```diff
@@ -1223,22 +1223,53 @@ def cross_entropy(input,
...
+                if (weight is None):
+                    count = core.ops.reduce_sum(mask, 'reduce_all', True)
+                    ret = out_sum / count
+                else:
+                    mask = paddle.cast(mask, "float64")
+                    weight_gather_reshape = paddle.cast(weight_gather_reshape,
+                                                        "float64")
```
Comment (willthefrog) [inline python/paddle/nn/functional/loss.py]: why hard code to "float64"?

File: python/paddle/nn/functional/loss.py [hunk 5]
```diff
@@ -1223,22 +1223,53 @@ def cross_entropy(input,
...
+                    weight_ignored = core.ops.elementwise_mul(
```
Comment (willthefrog) [inline python/paddle/nn/functional/loss.py:1256]: also, the logic could use some simplification, maybe move the `ignore_index` condition to the top? so the math in general become something like:
```python
weight_term = mask_in_float * weight_gathered
sum(out * weight_term) / sum(weight_term) 
```

File: python/paddle/nn/functional/loss.py [hunk 6]
```diff
@@ -1258,20 +1285,41 @@ def cross_entropy(input,
...
     elif reduction == "mean":
-        if weight is not None:
+        if ignore_index != -100:
+            out_sum = paddle.sum(out, name=name)
+            #for each label[i],set 1 or 0, according to ignore_index
+            ones_t = paddle.ones_like(label)
+            zeros_t = paddle.zeros_like(label)
+            #mask[i]=0, if label[i]==ignore_index
+            #mask[i]=1, otherwise 
+            mask = paddle.where(label != ignore_index, ones_t, zeros_t)
```
Comment (willthefrog) [inline python/paddle/nn/functional/loss.py]: use the same mask generation as above?

Comment (chajchaj) [reply python/paddle/nn/functional/loss.py]: ok

File: python/paddle/nn/functional/loss.py [hunk 7]
```diff
@@ -1258,20 +1285,37 @@ def cross_entropy(input,
...
     if reduction == "sum":
         return paddle.sum(out, name=name)
     elif reduction == "mean":
-        if weight is not None:
+        if ignore_index != -100:
+            out_sum = paddle.sum(out, name=name)
+            #for each label[i],set 1 or 0, according to ignore_index
+            #mask[i]=0, if label[i]==ignore_index
+            #mask[i]=1, otherwise 
+            mask = paddle.cast(label != ignore_index, dtype='int32')
```
Comment (willthefrog) [inline python/paddle/nn/functional/loss.py]: no need to be int32, just use same dtype as weight? and the cast below can be removed as a result.

**Review by willthefrog — APPROVED** [review]
LGTM

**Pull request merged**